File: Leetcode/543. Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int mx = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return mx-1;
    }
    
    int dfs(TreeNode* root) {
        if(!root) return 0;
        
        int l = dfs(root->left);
        int r = dfs(root->right);
        
        mx = max(mx, l+r+1);
        
        return max(l,r) + 1;
    }
};

File: Leetcode/TypeScript/543. Diameter of Binary Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
let ans = 0;
function diameterOfBinaryTree(root: TreeNode | null): number {
    ans = 0;
    dfs(root);
    return ans-1;
};

const dfs = (root: TreeNode | null): number => {
    if(!root) return 0;
    let l = dfs(root.left);
    let r = dfs(root.right);
    ans = Math.max(ans, l+r+1);
    return Math.max(l,r) +1;
}
